### DISCUSSION TITLE
Reaparr Integration ideas

### DISCUSSION THREAD

**Discussion by JasonLandbridge** [PlexRipper/PlexRipper#400] [Ideas]
**EDIT: If you're looking for the current state of Sonarr/Radarr integration then go here: https://www.reaparr.rocks/docs/integration/sonarr-radarr**

Hi there!
 
Please post below what type of integrations you would like Reaparr to have and how it should work. Think of, for example, Sonarr/Radarr integrations, whereby media requested in those are sent to Reaparr and be downloaded. Or something with Overseer or anything else you can come with.

The more details, the better, and I will add it to the plan below!

# Features

## Radarr/Sonarr

NOTE: From Discord chats in the [Servarr Discord](https://discord.com/invite/8Dbsx35rrx)</summary>, it seems there is a great dislike towards Reaparr, see screenshots. This means I won't expect any work on their side. Fortunately, there is a well documented API with which we can work, without any changes on their side. This is why we have to let Reaparr work as a Torznab indexer and a torrent download client

<details>
<summary>Screenshots</summary>

![image](https://github.com/user-attachments/assets/e3bb0859-29a1-409e-9b31-edffe10a6eef)
![image](https://github.com/user-attachments/assets/c47acd8b-de04-4620-ba17-6c6919645ae1)
![image](https://github.com/user-attachments/assets/43e3aca3-869f-4fa8-aae7-a73d6a554f2b)

</details>

### Terms: 
-  [Sonarr](https://sonarr.tv/)
-  [Radarr](https://radarr.video/)
-  [Indexer](https://wiki.servarr.com/sonarr/settings#indexers)
-  [Download Client](https://wiki.servarr.com/sonarr/settings#download-clients)

### User stories
  - **Use Reaparr as a search indexer and download client, this requires**
	 - Implement Torznab endpoints for acting like an indexer according to [Torznab specification](https://torznab.github.io/spec-1.3-draft/torznab/Specification-v1.3.html)
       - Require API Token authentication on endpoints
       - Search endpoint with possible filters, all data should be retrievable from Plex API
	 - Implement download client as if qBittorrent. qBittorrent is chosen as it has it's own [well documented API](https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-(qBittorrent-4.1))
       - Require API Token authentication on endpoints
       - This includes communicating [the downloadstatus](https://github.com/Sonarr/Sonarr/blob/ab49268bac95949fd0b083e0c4357dd918de43f1/src/NzbDrone.Core/Download/Clients/QBittorrent/QBittorrent.cs#L247) as if it was qBittorrent
       - Implement automated clean-up once the file has been imported by Sonarr/Radarr
	 - Requires setting up Reaparr as a download client first, then as an Indexer in Sonarr and force the indexer to always use the Reaparr download Client. This last one is important, as the response from the Reaparr indexer is a fake torrent, which only the Reaparr download client will understand. This is why Reaparr (possibly) cannot be used with something like [Prowlarr](https://prowlarr.com/), see [Download Clients (Prowlarr Searches)](https://wiki.servarr.com/prowlarr/quick-start-guide#download-clients-prowlarr-searches)
		 <details>
		 <summary>Setting Reaparr as the download client for an indexer</summary>

		 ![image](https://github.com/user-attachments/assets/ba158a43-1148-4d8c-8ee4-106fe442dc88)

		 ![image](https://github.com/user-attachments/assets/4d07cade-3b31-4a86-9cc6-a17a8e131dc9)

		 </details>
	 - Downloaded files are put into a folder, which are then moved and imported by Sonarr/Radarr
  - **Push Plex Library list to Sonarr for downloading there** 
    - Requested here: [#264](https://github.com/Reaparr/Reaparr/issues/264)
    - Per Plex library, it could be possible to push all media as a list to Sonarr/Radarr through the use of [Lists](https://wiki.servarr.com/sonarr/settings#lists). Reaparr will not download the media files but merely convert it into a list of media and push it to Sonarr/Radarr.
    - Requires adding Sonarr/Radarr connection instance to Reaparr to which to send the list 

## Overseer

Note: I have no experience with Overseer so the below has to be worked out more after more investigation. At least Overseer is seperate from the above *arr group so thighter integration might be on the table. Regardless, the Sonarr/Radarr has a higher priority of implementation

### Terms: 
-  [Overseer](https://overseerr.dev/)

### User stories
 -   **Notify of recently added Plex media to Overseer such that it can be requested**
 -   **Create its own "wanted" system of media that is synced with Overseer**
 -   **(Periodic?) check with Reaparr if a certain media has been found, and then download it after approval/automatically?**

**Comment by D00MSDAYDEVICE** [comment]
A notification when something in your Overseer request list appears in a shared library.

**Reply by JasonLandbridge** [reply]
I've never used [Overseerr](https://overseerr.dev/), but from what I gather it could work in the following ways (add more ideas):

 - **PlexRipper -> Overseer:**
   - Notify of recently added Plex media to Overseer such that it can be requested
   - Create its own "wanted" system of media that is synced with Overseer
- **Overseer -> PlexRipper:**
  - (Periodic?) check with PlexRipper if a certain media has been found, and then download it after approval/automatically? 

This is just from the top of my head, but let me know if this is what you mean.

**Reply by D00MSDAYDEVICE** [reply]
To clarify, Overseerr has it's own existing request list. Usually Overseerr periodically looks for these requests via sonarr/radarr.  If a shared media library the client has access to acquired something from this request list, it would be good to get a notification it's available (I personally would not setup auto-downloads). This function might be more of an integration for Overseerr (you'd need to add Plexripper as a server under services), but if there's any method to "pull" the request list from it, then compare that list to your shared libraries, it would be beneficial.

Currently, there are multiple methods in Overseerr for notifications (Settings - Notifications - Webpush/Discord/Webhook, etc). These include:

* Request Pending Approval
Send notifications when users submit new media requests which require approval.

* Request Automatically Approved
Send notifications when users submit new media requests which are automatically approved.

If this could be leveraged I think it would be useful. Comparing and scraping could be done on the Plexripper side.

TLDR: The second method with a notification something new was found.

**Comment by petebutty** [comment]
Sonarr and radarr support should be fantastic, implementation with nzb360 through API would also be my no. 1 ask

**Reply by JasonLandbridge** [reply]
What would you like to see in NZB360? Download progress? Or pick what you want to download through PlexRipper? Can you be more specific?

**Reply by petebutty** [reply]
Looking through the recent messages, I think adding it as a prowlarr indexer would be great, nzb360 already had implementation to add it as a service which opens the web page.

**Reply by JasonLandbridge** [reply]
I had updated the opening post, see this part:

> Requires setting up PlexRipper as a download client first, then as an Indexer in Sonarr and force the indexer to always use the PlexRipper download Client. This last one is important, as the response from the PlexRipper indexer is a fake torrent, which only the PlexRipper download client will understand. This is why PlexRipper (possibly) cannot be used with something like [Prowlarr](https://prowlarr.com/), see [Download Clients (Prowlarr Searches)](https://wiki.servarr.com/prowlarr/quick-start-guide#download-clients-prowlarr-searches)

So Prowlarr won't work, NZB360 is an option but again, what do you want to see when using it with PlexRipper? Download stats? etc

**Comment by silkyclouds** [comment]
**To integrate PlexRipper with the “Arrs” ecosystem (Prowlarr, Radarr, Sonarr, Lidarr) in two key ways:**

**Step 1: Add PlexRipper as an Indexer**

Goal: Enable PlexRipper to function as an indexer within the Arrs (at Prowlarr or lower levels like Radarr/Sonarr/Lidarr).

**Outcome:**

- 	Users can search for content indexed by PlexRipper directly from the Arrs interfaces.
- 	Selected content can then be sent to the appropriate Arr-managed download client for processing.

**Step 2: Modify PlexRipper to Function as a Download Client**

Goal: Allow PlexRipper to emulate a Usenet or torrent download client.

**Approach:**

- 	Reuse existing settings such as URL, username, password, etc., to connect Arrs to PlexRipper instead of a traditional client like qBittorrent.

Develop a backend for PlexRipper capable of:
- 	Receiving search inputs from Arrs.
- 	Identifying and selecting preferred media versions based on user-defined filters within PlexRipper.
- 	Starting the download process directly within PlexRipper.

**Challenges:**

PlexRipper must manage additional responsibilities like:
- 	Handling search and download tasks based on Arrs input.
- 	Applying user preferences (e.g., quality, language filters).

**Limitations:**
- 	Certain features won’t be supported initially (e.g., reporting download completion to Arrs, automated cleanup of completed downloads).
- 	Heavy Development Required: Significant backend work is needed to ensure compatibility with Arr workflows and user expectations.

**Questions for You:**

- 	Are you currently experimenting with interfacing PlexRipper and the Arrs?
- 	Or are you still brainstorming and planning the integration approach?

**Reply by JasonLandbridge** [reply]
I can see that you're a software developer as you're the first one who wrote so much needed detail, thanks for that! :smile: 

> Step 1: Add PlexRipper as an Indexer
> Step 2: Modify PlexRipper to Function as a Download Client

I like the idea of setting up the PlexRipper API as if it is an Usenet indexer and an Usenet Download Client. This would avoid a lot of custom implementation in the various *Arrs programs. There is a big obstacle in this, it would be a "fake" indexer and any results returned by PlexRipper would only work with PlexRipper and not with any other download client. It seems as if Sonarr/Radarr will ask all indexers for a search result, and then send them to any of the registered download clients. Which will not work as the PlexRipper search result doesn't exist on any Usenet server. 

Maybe a compromise could be that any PlexRipper indexer results will only work with a PlexRipper download client, this should be relatively "easy".  After a PlexRipper download is completed, Sonarr/Radarr could then move the download to its final destination as currently happens with Usenet/Torrent downloads. 

> Challenges:
> PlexRipper must manage additional responsibilities like:
> - Handling search and download tasks based on Arrs input.
> - Applying user preferences (e.g., quality, language filters).

Yes, good points! I think most, if not all, data points should be available in the Plex data that is synced with PlexRipper, and otherwise I will start storing it. 


> Limitations:
> - Certain features won’t be supported initially (e.g., reporting download completion to Arrs, automated cleanup of completed downloads).

Fortunately, Both PlexRipper and Sonarr/Radarr use the same .NET backend with something called SignalR. Which would allow for real-time progress reporting. Automated clean-up is something that can be done automatically in PlexRipper, something like remove downloads after x days. 

> - Heavy Development Required: Significant backend work is needed to ensure compatibility with Arr workflows and user expectations.

Very true! Which is why I would like to plan this out as much as possible to avoid wasting development time and getting to know what users would like. I will reach out to *arr developers when we have a semi-plan/proposal here and get some advice. 


> Questions for You:

> Are you currently experimenting with interfacing PlexRipper and the Arrs?

Nothing concrete yet, but it's the most requested feature and has been one of my goals since starting PlexRipper. 

>Or are you still brainstorming and planning the integration approach?

Mostly this, I'm thorn between making a custom API where I have the freedom to add support for any extra future PlexRipper feature. But this is quite dangerous, as all *arrs need to also accommodate this freedom. Or play it safe by going to Torrent/Usenet route, but I'm not sure if this "fits" the way PlexRipper works. 

Maybe I will start with a list of features and the user flow to get an idea what is needed.

**Reply by JasonLandbridge** [reply]
@silkyclouds You were spot on! I've looked into it more, and your proposal seems to be the most viable path forward, please read the opening post with the updated plan. I would love to know what you think about it and how I can improve it!

**Comment by jlightner86** [comment]
Hello!

First, let me say I love your work. 

A couple suggestions:

First - Have an option to add a checkmark to the posters when you have downloaded something. 

Second - Have a setting symbol next to each folder, so you can select a default folder to download/merge into. It gets very repetitive clicking the download link, then clicking the arrow next to the download button, and then selecting a different folder for every single item. 

Third - Have an option to select multiple items and then add them to download queue or click on the library and have the option to download the entire library. This would be great for small libraries, or libraries that may not exist yet on the destination machine.

**Comment by aniel300** [comment]
any update on this?

**Comment by baxterblk** [comment]
I'm so sorry I couldn't return to this, Jason! Glad to see you still found a way to make it happen. Well done, sir!

**Comment by SammyTheKat** [comment]
With the recent name change to Reaparr are we closer to some sort of Sonarr/Radarr integration? Amazing project!

**Comment by alvarodevlabs** [comment]
Still waiting this feature! Hope you can update us with some little new info!

**Reply by JasonLandbridge** [reply]
Very soon, you can however already try the experimental version on the dev docker image. Also works on ARM cpu now

**Reply by JoelyMoley** [reply]
> Very soon, you can however already try the experimental version on the dev docker image. Also works on ARM cpu now

assuming this is in regards to importing shared library movies/tv shows directly into sonarr/radarr? playing with the latest dev release and still can't see how this is done, thanks mate!

**Reply by JasonLandbridge** [reply]
@JoelyMoley See: https://www.reaparr.rocks/docs/integration/sonarr-radarr

That page describes what the current live functionalities are
